com.wordnik.swagger.jaxrs.utils.ParameterUtils does not support Nested ParameterizedTypes

If I user nested ParameterizedTypes (E.g RequestEntity<Pageable<Unqualified, PartnerPublicFilter>> query) in the ApiOperation method I get the following exception:

[ERROR] Failed to execute goal com.github.kongchen:swagger-maven-plugin:3.0.0:generate (default-cli) on project pass-mepo-server: sun.reflect.generics.reflectiveObjects.ParameterizedTypeImpl cannot be cast to java.lang.Class -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.github.kongchen:swagger-maven-plugin:3.0.0:generate (default-cli) on project pass-mepo-server: sun.reflect.generics.reflectiveObjects.ParameterizedTypeImpl cannot be cast to java.lang.Class
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:216)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:108)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:76)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:116)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:361)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:155)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:213)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:157)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:483)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: sun.reflect.generics.reflectiveObjects.ParameterizedTypeImpl cannot be cast to java.lang.Class
        at com.github.kongchen.swagger.docgen.mavenplugin.ApiDocumentMojo.execute(ApiDocumentMojo.java:91)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:133)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        ... 19 more
Caused by: java.lang.ClassCastException: sun.reflect.generics.reflectiveObjects.ParameterizedTypeImpl cannot be cast to java.lang.Class
        at com.wordnik.swagger.jaxrs.utils.ParameterUtils.isMethodArgumentAnArray(ParameterUtils.java:43)
        at com.github.kongchen.swagger.docgen.reader.AbstractReader.getParameters(AbstractReader.java:275)
        at com.github.kongchen.swagger.docgen.reader.JaxrsReader.parseMethod(JaxrsReader.java:345)
        at com.github.kongchen.swagger.docgen.reader.JaxrsReader.read(JaxrsReader.java:106)
        at com.github.kongchen.swagger.docgen.reader.JaxrsReader.read(JaxrsReader.java:66)
        at com.github.kongchen.swagger.docgen.reader.JaxrsReader.read(JaxrsReader.java:57)
        at com.github.kongchen.swagger.docgen.mavenplugin.MavenDocumentSource.loadDocuments(MavenDocumentSource.java:40)
        at com.github.kongchen.swagger.docgen.mavenplugin.ApiDocumentMojo.execute(ApiDocumentMojo.java:69)
        ... 21 more

--- a/modules/swagger-core/src/main/java/com/wordnik/swagger/jackson/ModelResolver.java
+++ b/modules/swagger-core/src/main/java/com/wordnik/swagger/jackson/ModelResolver.java
@@ -245,19 +245,6 @@ public class ModelResolver extends AbstractModelConverter implements ModelConver
       model.xml(xml);
     }
 
-    ApiModel apiModel = beanDesc.getClassAnnotations().get(ApiModel.class);
-    // TODO
-    if (apiModel != null) {
-      if(apiModel.value() != null && !"".equals(apiModel.value())) {
-        name = apiModel.value();
-        model.setName(name);
-      }
-      Class<?> parent = apiModel.parent();
-      if (parent != Void.class) {
-        // model.setBaseModel(_typeName(_mapper.constructType(parent)));
-      }
-    }
-
     // see if @JsonIgnoreProperties exist
     Set<String> propertiesToIgnore = new HashSet<String>();
     JsonIgnoreProperties ignoreProperties = beanDesc.getClassAnnotations().get(JsonIgnoreProperties.class);
@@ -265,6 +252,7 @@ public class ModelResolver extends AbstractModelConverter implements ModelConver
       propertiesToIgnore.addAll(Arrays.asList(ignoreProperties.value()));
     }
 
+    final ApiModel apiModel = beanDesc.getClassAnnotations().get(ApiModel.class);
     String disc = (apiModel == null) ? "" : apiModel.discriminator();
     if (disc.isEmpty()) {
       // longer method would involve AnnotationIntrospector.findTypeResolver(...) but:

--- a/modules/swagger-core/src/main/java/com/wordnik/swagger/jackson/TypeNameResolver.java
+++ b/modules/swagger-core/src/main/java/com/wordnik/swagger/jackson/TypeNameResolver.java
@@ -6,7 +6,11 @@ import java.net.URI;
 import java.net.URL;
 import java.util.*;
 
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.text.WordUtils;
+
 import com.fasterxml.jackson.databind.JavaType;
+import com.wordnik.swagger.annotations.ApiModel;
 
 /**
  * Helper class used for converting well-known (property) types into
@@ -38,9 +42,31 @@ public class TypeNameResolver {
   protected TypeNameResolver() { }
 
   public String nameForType(JavaType type) {
+    if (type.hasGenericTypes()) {
+      return nameForGenericType(type);
+    }
     final Class<?> raw = type.getRawClass();
-    String name = findStdName(raw);
-    return (name == null) ? name = raw.getSimpleName() : name;
+    final String name = findStdName(raw);
+    return (name == null) ? nameForClass(raw) : name;
+  }
+
+
+  protected String nameForClass(Class<?> cls) {
+    final ApiModel model = cls.getAnnotation(ApiModel.class);
+    final String modelName = model == null ? null : StringUtils.trimToNull(model.value());
+    return modelName == null ? cls.getSimpleName() : modelName;
+  }
+
+  protected String nameForGenericType(JavaType type) {
+    final StringBuilder generic = new StringBuilder(nameForClass(type.getRawClass()));
+    final int count = type.containedTypeCount();
+    for (int i = 0; i < count; ++i) {
+      final JavaType arg = type.containedType(i);
+      final Class<?> argClass = arg.getRawClass();
+      final String argName = findStdName(argClass) != null ? nameForClass(argClass) : nameForType(arg);
+      generic.append(WordUtils.capitalize(argName));
+    }
+    return generic.toString();
   }
 
   protected String findStdName(Class<?> raw) {
@@ -90,7 +116,7 @@ public class TypeNameResolver {
     map.put("javax.xml.datatype.XMLGregorianCalendar", TYPE_DATE_TIME);
     return map;
   }
-  
+
   private static Map<Class<?>, String> _add(Map<Class<?>, String> map, String name,
       Class<?>... types) {
     for (Class<?> type : types) {

--- a/modules/swagger-jaxrs/src/test/scala/GenericsTest.scala
+++ b/modules/swagger-jaxrs/src/test/scala/GenericsTest.scala
@@ -1,4 +1,5 @@
 import scala.collection.JavaConverters.asScalaBufferConverter
+import scala.collection.JavaConverters.asScalaSetConverter
 
 import org.junit.runner.RunWith
 import org.scalatest.FlatSpec
@@ -7,7 +8,9 @@ import org.scalatest.junit.JUnitRunner
 
 import com.wordnik.swagger.jaxrs.Reader
 import com.wordnik.swagger.models.ArrayModel
+import com.wordnik.swagger.models.ModelImpl
 import com.wordnik.swagger.models.Operation
+import com.wordnik.swagger.models.RefModel
 import com.wordnik.swagger.models.Swagger
 import com.wordnik.swagger.models.parameters.BodyParameter
 import com.wordnik.swagger.models.parameters.QueryParameter
@@ -49,6 +52,10 @@ class GenericsTest extends FlatSpec with Matchers {
     p.getItems should be (null)
   }
 
+  def testGenericType(op: Operation, `type`: String) = {
+    getBodyParameter(op, 0).getSchema.asInstanceOf[RefModel].getSimpleRef should be (`type`)
+  }
+
   def getOperation(name: String) = {
     swagger.getPath("/generics/" + name).getPost
   }
@@ -59,6 +66,10 @@ class GenericsTest extends FlatSpec with Matchers {
 
   def getBodyParameter(op: Operation, index: Int) = {
     op.getParameters.get(index).asInstanceOf[BodyParameter]
+  }
+
+  def getProperty(`type`: String, name: String) = {
+    swagger.getDefinitions.get(`type`).asInstanceOf[ModelImpl].getProperties.get(name)
   }
 
   it should "check collections of integers" in {
@@ -138,5 +149,38 @@ class GenericsTest extends FlatSpec with Matchers {
     val p = getBodyParameter(op, 0)
     val ddArray = p.getSchema.asInstanceOf[ArrayModel]
     ddArray.getItems.asInstanceOf[ArrayProperty].getItems.asInstanceOf[RefProperty].getSimpleRef should be ("Tag")
+  }
+
+  it should "check parameters of generic types" in {
+    val genericTypes = Set("GenericTypeString", "GenericTypeUUID", "GenericTypeGenericTypeString",
+        "RenamedGenericTypeString", "RenamedGenericTypeRenamedGenericTypeString")
+    (swagger.getDefinitions.keySet.asScala & genericTypes) should be (genericTypes)
+
+    val opString = getOperation("testGenericType")
+    testGenericType(opString, "GenericTypeString")
+    val strValue = getProperty("GenericTypeString", "value")
+    strValue should not be (null)
+    strValue.getClass().getName() should be (classOf[StringProperty].getName)
+
+    val opUUID = getOperation("testStringBasedGenericType")
+    testGenericType(opUUID, "GenericTypeUUID")
+    val uuidValue = getProperty("GenericTypeUUID", "value")
+    uuidValue should not be (null)
+    // TODO
+    //uuidValue.getClass().getName() should be (classOf[UUIDProperty].getName)
+
+    val opComplex = getOperation("testComplexGenericType")
+    testGenericType(opComplex, "GenericTypeGenericTypeString")
+    val complexValue = getProperty("GenericTypeGenericTypeString", "value")
+    complexValue should not be (null)
+    complexValue.getClass().getName() should be (classOf[RefProperty].getName)
+    complexValue.asInstanceOf[RefProperty].getSimpleRef should be ("GenericTypeString")
+
+    val opRenamed = getOperation("testRenamedGenericType")
+    testGenericType(opRenamed, "RenamedGenericTypeRenamedGenericTypeString")
+    val renamedComplexValue = getProperty("RenamedGenericTypeRenamedGenericTypeString", "value")
+    renamedComplexValue should not be (null)
+    renamedComplexValue.isInstanceOf[RefProperty] should be (true)
+    renamedComplexValue.asInstanceOf[RefProperty].getSimpleRef should be ("RenamedGenericTypeString")
   }
 }

--- a/modules/swagger-jaxrs/src/test/scala/models/GenericType.java
+++ b/modules/swagger-jaxrs/src/test/scala/models/GenericType.java
@@ -1,0 +1,5 @@
+package models;
+
+public class GenericType<T> {
+  public T value;
+}

--- a/modules/swagger-jaxrs/src/test/scala/models/GenericTypeWithApiModel.java
+++ b/modules/swagger-jaxrs/src/test/scala/models/GenericTypeWithApiModel.java
@@ -1,0 +1,8 @@
+package models;
+
+import com.wordnik.swagger.annotations.ApiModel;
+
+@ApiModel("RenamedGenericType")
+public class GenericTypeWithApiModel<T> {
+  public T value;
+}

--- a/modules/swagger-jaxrs/src/test/scala/resources/ResourceWithGenerics.java
+++ b/modules/swagger-jaxrs/src/test/scala/resources/ResourceWithGenerics.java
@@ -2,11 +2,14 @@ package resources;
 
 import java.util.List;
 import java.util.Set;
+import java.util.UUID;
 
 import javax.ws.rs.POST;
 import javax.ws.rs.Path;
 import javax.ws.rs.QueryParam;
 
+import models.GenericType;
+import models.GenericTypeWithApiModel;
 import models.Tag;
 import models.TestEnum;
 
@@ -73,4 +76,29 @@ public class ResourceWithGenerics {
   @ApiOperation("Tests object container as body parameter")
   public void test2DInBody(List<List<Tag>> list) {
   }
+
+  @ApiOperation(value="Tests generic type")
+  @POST
+  @Path("/testGenericType")
+  public void testGenericType(GenericType<String> type) {
+  }
+
+  @ApiOperation(value="Tests generic type")
+  @POST
+  @Path("/testStringBasedGenericType")
+  public void testStringBasedGenericType(GenericType<UUID> type) {
+  }
+
+  @ApiOperation(value="Tests complex generic type")
+  @POST
+  @Path("/testComplexGenericType")
+  public void testComplexGenericType(GenericType<GenericType<String>> type) {
+  }
+
+  @ApiOperation(value="Tests renamed generic type")
+  @POST
+  @Path("/testRenamedGenericType")
+  public void testRenamedGenericType(GenericTypeWithApiModel<GenericTypeWithApiModel<String>> type) {
+  }
+
 }

